File: android/app/src/main/kotlin/ca/lucianlabs/tgsocial/ui/components/PostHeader.kt
package ca.lucianlabs.tgsocial.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.unit.Dp
import ca.lucianlabs.housepour.HPAvatar
import ca.lucianlabs.housepour.HPBody
import ca.lucianlabs.housepour.HPButton
import ca.lucianlabs.housepour.HPButtonSize
import ca.lucianlabs.housepour.HPButtonStyle
import ca.lucianlabs.housepour.HPHitGrow
import ca.lucianlabs.housepour.HPHitTarget
import ca.lucianlabs.housepour.HPMonoSmall
import ca.lucianlabs.housepour.HPTokens
import ca.lucianlabs.housepour.hpHitBandBelow
import ca.lucianlabs.housepour.hpLineBox

/**
 * The gap a post card holds between [PostHeader] and the first **clickable** thing under it.
 *
 * `rowGap` is the rhythm; rule 6 is the floor. The channel subheading is one mono-small line box (19.2dp) and
 * takes the rest of its `touchMin` as an overlay hanging below itself — so the 20.8dp under the header belongs
 * to the channel, not to the card. Compose hit-tests later-placed siblings first, so a clickable sibling that
 * starts inside that band swallows it: with `rowGap` alone the channel measured 40dp and lived at 30dp, and the
 * bottom 10dp of it opened the thread. Holding the band makes the boundary between the channel and the body a
 * line, exactly the way the name and the channel already tile with each other.
 */
val PostHeaderBottomGap: Dp = maxOf(HPTokens.Space.rowGap, hpHitBandBelow(HPTokens.Type.monoSmall.hpLineBox))

/**
 * PRODUCT §2.3 — the post header: the source-channel avatar, the name/channel stack, then the time and Share.
 *
 * The metrics are the spec's: the stack is **tight** (the name at the body line height, the channel directly
 * under it at the mono-small line height, no extra leading), the avatar is **centred against that stack**
 * rather than pinned to the top of it, and the whole row measures about one avatar tall.
 *
 * Every control keeps its `touchMin` hit target as an **overlay** ([HPHitTarget]) — nothing here is padded out
 * to 40dp and pulled back with a negative margin, which is what leaves a 47pt box around 19pt of text. The two
 * stack rows grow away from each other so the split between their targets lands exactly on the boundary
 * between their line boxes.
 *
 * The channel's half of that split hangs below the header, so the header does not own all of its own hit
 * targets: the card has to keep [PostHeaderBottomGap] clear underneath. See that value for why.
 *
 * Split out of [PostCard] so these metrics can be measured directly (see `PostHeaderTest` for what they
 * measure and `PostCardHitRegionTest` for what they do once the card puts something under them).
 */
@Composable
fun PostHeader(
    avatar: ImageBitmap?,
    name: String,
    initial: String,
    channelTitle: String?,
    time: String,
    onOpenName: () -> Unit,
    onOpenChannel: () -> Unit,
    onShare: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier,
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(HPTokens.Space.rowGap),
    ) {
        // The avatar is the source channel (§2.3); tapping it goes where the name goes.
        HPHitTarget(onClick = onOpenName, contentDescription = name) {
            HPAvatar(avatar, HPTokens.Space.avatarRow, initial)
        }
        Column(modifier = Modifier.weight(1f)) {
            // The name is the node (tap → node profile); on the channel fallback it is the channel title.
            HPHitTarget(onClick = onOpenName, contentDescription = "Open $name", grow = HPHitGrow.UP) {
                HPBody(name, strong = true, maxLines = 1)
            }
            if (channelTitle != null) {
                // Subheading = the channel title, mono small muted, tap → feed channel screen (§2.6).
                HPHitTarget(onClick = onOpenChannel, contentDescription = "Open $channelTitle", grow = HPHitGrow.DOWN) {
                    HPMonoSmall(channelTitle, maxLines = 1)
                }
            }
        }
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(HPTokens.Space.rowGap)) {
            HPMonoSmall(time, color = HPTokens.Colors.faint, maxLines = 1)
            HPButton("Share", onShare, style = HPButtonStyle.GHOST, size = HPButtonSize.SMALL, contentDescription = "Share")
        }
    }
}

File: ios/Sources/Components/PostHeader.swift
// Components — the post card header (PRODUCT.md §2.3 "Header metrics").
//
// One row: the avatar, then the name/channel stack, then the time and Share. The stack is *tight* —
// the name at the body line height, the channel directly under it at the mono-small line height, no
// extra leading — and the avatar is centred against that stack, not pinned above it. The whole
// header measures about one avatar tall; a header appreciably taller than its own avatar means
// something in it has been inflated.
//
// So nothing here reaches its hit target (COMPONENTS.md rule 6) by growing a line box. Every control
// wears a clear overlay instead, one that extends past the painted bounds and leaves the laid-out
// height alone. The two stacked labels anchor their overlays to the edge they share, so the name
// grows its region upwards and the channel grows its own downwards and neither covers the other's
// glyphs.
//
// A region is only real where the space around it is clear, and the channel's half hangs *below the
// header*: the header does not own all of its own hit targets. The card owes it `PostHeaderBottomGap`
// clear of anything tappable, exactly as on Android (`PostHeader.kt`, `PostHeaderBottomGap`) — and
// without that band the 40pt overlay is a comment rather than a target, because the post body is laid
// out after the header and takes every point the two share. `PostHeaderHitRegionTests` measures the
// regions inside a real card and fails if any of them reaches into a neighbour's tap surface.
//
// Generic over the avatar so `PostHeaderMetricsTests` can measure the real layout without an
// AppModel; the app always passes `NodeAvatar`.

import SwiftUI

/// The band the card holds clear of tap surfaces **under** the header (COMPONENTS.md rule 6, the
/// tiling half). The channel subheading is one mono-small line box and takes the rest of its
/// `touchMin` as an overlay hanging below itself, so that much of the card's next gap belongs to the
/// channel, not to whatever follows: SwiftUI hit-tests later siblings first, and one that starts
/// inside the band swallows it — the target measures 40pt and lives at 14pt, its bottom 26pt opening
/// the thread instead of the feed. `rowGap` is the rhythm; rule 6 is the floor. Mirrors Android's
/// `PostHeaderBottomGap` (there the ramp's explicit line height makes it 20.8dp; SwiftUI paints a
/// shorter line box, so the overhang — and the band — is larger here).
let PostHeaderBottomGap: CGFloat = max(HPTokens.Space.rowGap, hpHitBandBelow(HPType.monoSmall.hpLineBox))

extension View {
    /// Holds `PostHeaderBottomGap` clear under a post header, less the `rowGap` the element below it
    /// carries as its own whitespace — the smallest gap any of them carries, so every card shape
    /// (text at `rowPad`, media and the footer at `rowGap`) keeps at least the band. Whitespace
    /// counts: what the band may not contain is a *tap surface*, which is why `PostTextBlock` keeps
    /// its padding outside its content shape.
    ///
    /// One modifier so the card and `PostHeaderHitRegionTests` cannot drift: the test asserts the
    /// band on the same views the card ships, arranged the way the card arranges them.
    func postHeaderBottomBand() -> some View {
        padding(.bottom, max(0, PostHeaderBottomGap - HPTokens.Space.rowGap))
    }
}

/// Labels the post card's hit regions report under `hpMeasureTouchTargets` (`HPTouchProbe`), so a
/// test names a region instead of counting tree order.
enum PostCardRegion {
    static let avatar = "avatar"
    static let name = "name"
    static let channel = "channel"
    static let share = "share"
    static let text = "post text"
    /// The audio player's spectrogram strip (PRODUCT §2.11.1). Its painted shape is taller than
    /// `touchMin`, so unlike the header's controls it needs no overlay — but it still reports, so
    /// the assembled-card test can prove the region is really there.
    static let strip = "spectrogram strip"
}

struct PostHeader<Avatar: View>: View {
    /// The person — the node the post reaches me through, or the channel title when unattributed.
    let name: String
    /// The source channel's title. Nil when nothing attributes the post and `name` *is* the channel.
    let channel: String?
    let date: Int
    let shareURL: URL?
    /// PRODUCT §2.22.3: in the demo `Share` stays where it is, stays tappable, and answers rather
    /// than presenting a share sheet — a system sheet would put an invented `t.me` link into
    /// someone's messages. Non-nil replaces the `ShareLink` with a button that runs this.
    let onShareRefused: (() -> Void)?
    let onOpenName: () -> Void
    let onOpenChannel: () -> Void
    let avatar: Avatar

    init(name: String, channel: String?, date: Int, shareURL: URL?,
         onShareRefused: (() -> Void)? = nil,
         onOpenName: @escaping () -> Void, onOpenChannel: @escaping () -> Void,
         @ViewBuilder avatar: () -> Avatar) {
        self.name = name; self.channel = channel; self.date = date; self.shareURL = shareURL
        self.onShareRefused = onShareRefused
        self.onOpenName = onOpenName; self.onOpenChannel = onOpenChannel; self.avatar = avatar()
    }

    var body: some View {
        HStack(alignment: .center, spacing: HPTokens.Space.rowGap) {
            // The avatar is the source channel (§2.3) and shares the name's destination, as on
            // Android (PostCard.kt). 36pt painted, 40pt tappable.
            Button(action: onOpenName) {
                avatar.hpTouchOverlay(label: PostCardRegion.avatar)
            }
            .buttonStyle(.plain)
            .accessibilityLabel(name)

            VStack(alignment: .leading, spacing: 0) {
                // The name is the person: body strong, tap → node profile.
                Button(action: onOpenName) {
                    HPBody(name, strong: true)
                        .lineLimit(1)
                        .multilineTextAlignment(.leading)
                        // Upwards, into the card's own `cardPad` — clear space, so the full
                        // `touchMin` is real here.
                        .hpTouchOverlay(.bottomLeading, label: PostCardRegion.name)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Open \(name)")

                // Subheading: the channel, mono small muted, tap → feed channel screen (§2.6).
                //
                // Downwards, past the header and into `PostHeaderBottomGap` — the band the card
                // holds clear for exactly this. See that constant.
                if let channel {
                    Button(action: onOpenChannel) {
                        HPMonoSmall(channel)
                            .lineLimit(1)
                            .hpTouchOverlay(.topLeading, label: PostCardRegion.channel)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Open \(channel)")
                }
            }

            Spacer(minLength: HPTokens.Space.rowGap)

            HStack(alignment: .center, spacing: HPTokens.Space.rowGap) {
                // Relative time, mono faint, refreshed each minute while visible. Not a control:
                // the exact timestamp lives in the long-press sheet.
                TimelineView(.everyMinute) { context in
                    HPMonoSmall(PostTime.relative(unix: date, now: context.date), color: HPTokens.Colors.faint)
                }
                shareButton
            }
            // The trailing group takes its ideal width first; a long name or channel title is what
            // gives, never the time.
            .layoutPriority(1)
        }
    }

    /// Share — ghost small button right of the time (§2.3): the system share sheet with the post's
    /// t.me link. Its own padding keeps it under the avatar's height, so it never sets the row.
    @ViewBuilder private var shareButton: some View {
        // Same painted control and the same hit region either way, so `PostHeaderHitRegionTests`
        // measures one shape and §2.22.3's "nothing greyed out, nothing hidden" holds literally.
        if let onShareRefused {
            Button(action: onShareRefused) { shareLabel }
                .buttonStyle(HPPressStyle())
                .accessibilityLabel("Share")
        } else if let shareURL {
            ShareLink(item: shareURL) { shareLabel }
                .buttonStyle(HPPressStyle())
                .accessibilityLabel("Share")
        }
    }

    private var shareLabel: some View {
        Text("Share")
            .hpStyle(HPType.buttonSm, color: HPTokens.Colors.muted)
            .lineLimit(1)
            .padding(.vertical, HPTokens.Space.buttonSmY)
            .padding(.horizontal, HPTokens.Space.buttonSmX)
            .hpTouchOverlay(label: PostCardRegion.share)
    }
}
